rllib/algorithms/dqn/dqn.py

```diff
@@ -370,62 +369,62 @@ def training_step(self) -> ResultDict:
-                self.local_replay_buffer,
-                self.config["train_batch_size"],
-                count_by_agent_steps=self._by_agent_steps,
-            )
-
-            # Old-style replay buffers return None if learning has not started
-            if train_batch is None or len(train_batch) == 0:
-                self.workers.local_worker().set_global_vars(global_vars)
-                break
-
-            # Postprocess batch before we learn on it
-            post_fn = self.config.get("before_learn_on_batch") or (lambda b, *a: b)
-            train_batch = post_fn(train_batch, self.workers, self.config)
-
-            # for policy_id, sample_batch in train_batch.policy_batches.items():
-            #     print(len(sample_batch["obs"]))
-            #     print(sample_batch.count)
-
-            # Learn on training batch.
-            # Use simple optimizer (only for multi-agent or tf-eager; all other
-            # cases should use the multi-GPU optimizer, even if only using 1 GPU)
-            if self.config.get("simple_optimizer") is True:
-                train_results = train_one_step(self, train_batch)
-            else:
-                train_results = multi_gpu_train_one_step(self, train_batch)
-
-            # Update replay buffer priorities.
-            update_priorities_in_replay_buffer(
-                self.local_replay_buffer,
-                self.config,
-                train_batch,
-                train_results,
-            )
+        # Update target network every `target_network_update_freq` sample steps.
+        cur_ts = self._counters[
+            NUM_AGENT_STEPS_SAMPLED if self._by_agent_steps else NUM_ENV_STEPS_SAMPLED
+        ]
+
+        if cur_ts > self.config["num_steps_sampled_before_learning_starts"]:
+            for _ in range(sample_and_train_weight):
+                # Sample training batch (MultiAgentBatch) from replay buffer.
+                train_batch = sample_min_n_steps_from_buffer(
+                    self.local_replay_buffer,
+                    self.config["train_batch_size"],
+                    count_by_agent_steps=self._by_agent_steps,
+                )
 
-            # Update target network every `target_network_update_freq` sample steps.
-            cur_ts = self._counters[
-                NUM_AGENT_STEPS_SAMPLED
-                if self._by_agent_steps
-                else NUM_ENV_STEPS_SAMPLED
-            ]
-            last_update = self._counters[LAST_TARGET_UPDATE_TS]
-            if cur_ts - last_update >= self.config["target_network_update_freq"]:
-                to_update = self.workers.local_worker().get_policies_to_train()
-                self.workers.local_worker().foreach_policy_to_train(
-                    lambda p, pid: pid in to_update and p.update_target()
+                # Old-style replay buffers return None if learning has not started
+                if train_batch is None or len(train_batch) == 0:
```

Comment:
Is this code still needed? Can train_batch still be empty when you have waited for enough samples to get collected?